### DISCUSSION TITLE
How should I dispose an application

### DISCUSSION THREAD

**Discussion by GDX64** [leptos-rs/leptos#1560] [Q&A]
When we mount an application with `mont_to` or `mount_to_body`, the context disposer is forgeted by the `mount_to` function. Why not returning this from the function? How am I supose to dispose an application without this?

**Comment by gbj** [comment]
Why do you need to dispose of the root of the application?

**Comment by GDX64** [comment]
I want to put some wasm things on my website just to showcase some things and experiment with wasm. But I use vueJS with my website and just wanted to show some leptos apps in some routes. I'm using SPA and wanted to avoid reloading the page if possible. Here is one example I have using leptos: https://glmachado.com/wasm-chart

**Reply by gbj** [reply]
I guess I'd recommend just making a custom mounting function. It's not very complicated and I don't think any of the APIs involved are private. I'm not sure how you'd best call that disposer from Vue when navigating away from the page.
https://github.com/leptos-rs/leptos/blob/18deb398ca5f1644a38e61c2a658c4e13788bc58/leptos_dom/src/lib.rs#L892-L905

**Reply by GDX64** [reply]
I tried but could not use the `get_mountable_node` method. It looks like it is only available inside his module.

**Reply by gbj** [reply]
Yeah fair enough. I would say "feel free to make a PR to expose the right method" but the reactive ownership system has changed now on main in preparation for 0.5 so there isn't actually a scope and disposer created in mount_to any more.

**Comment by liquidnya** [comment]
My use-case for this is to have a component inside a React application that is written in Rust.

For `0.5.0-beta2` I created the following code:
```rust
#[cfg(target_arch = "wasm32")]
pub fn mount_to<F, N>(parent: web_sys::HtmlElement, f: F) -> leptos::Disposer
where
    F: Fn() -> N + 'static,
    N: leptos::IntoView,
{
    leptos::create_effect(move |last_node|
            // only render once in case effect is run again
            last_node.unwrap_or_else(|| {
                let node = f().into_view();
                parent.append_child(&leptos_dom::Mountable::get_mountable_node(&node)).unwrap();
                // store node in effect, so disposing the effect will drop the node
                node
            }))
    // convert effect into Disposer to dispose it on drop
    .into()
}

```
with the dependencies:
```toml
leptos = { version="0.5.0-beta2", features = ["csr"] }
leptos_dom = { version="0.5.0-beta2", features = ["csr"] }
web-sys = { version = "0.3.64", features = ["HtmlElement"] }

```
Then I call this with:
```rust
let disposer = mount_to(parent, || view! { <App/> });
```
Whenever `disposer` is dropped, the `App` is disposed.
Note that the `App` is still in the DOM.

Then I can move the `disposer` into the JavaScript world by using `wasm-bindgen`:
```rust
use wasm_bindgen::prelude::*;

#[wasm_bindgen(js_name = "App")]
pub struct JsApp {
    disposer: Option<leptos::Disposer>,
}

impl From<leptos::Disposer> for JsApp {
    fn from(disposer: leptos::Disposer) -> Self {
        Self {
            disposer: Some(disposer),
        }
    }
}

#[wasm_bindgen(js_class = "App")]
impl JsApp {
    /// For rendering a view that will never be unmounted from the dom use this.
    pub fn leak(&mut self) {
        std::mem::forget(self.disposer.take());
    }
}
```
Using `JsApp::from(disposer)` and handing it back to JavaScript will give the following methods:
```ts
export class App {
  free(): void;
/**
* For rendering a view that will never be unmounted from the dom use this.
*/
  leak(): void;
}
```

Now I can use `app.free()` and whenever the cleanup function is called in React.
The cleanup functions are then being run in leptos as a result.

Edit: fixed implementation of `JsApp::leak`.
